polar-core/src/bindings.rs

```diff
@@ -43,13 +47,135 @@ pub struct BindingManager {
     bindings: BindingStack,
 }
 
+/// The `BindingManager` maintains associations between variables and values,
+/// and constraints.
+///
+/// A variable may be:
+/// - unbound
+/// - bound
+/// - constrained
+///
+/// Variables may also be bound together such that their values or constraints
+/// will be the same.
+///
+/// A binding is created with the `bind` method.
+///
+/// The constraints or value associated with a variable is retrieved with `variable_state`.
 impl BindingManager {
     pub fn new() -> Self {
         Self::default()
     }
 
     /// Bind `var` to `val`.
+    ///
+    ///
+    /// If `var` is already bound or constrained, the
+    /// binding or constraints are replaced with `val`.
     pub fn bind(&mut self, var: &Symbol, val: Term) {
+        // TODO (dhatch): Would like to disable rebinding, but this has a large fallout.
+        // We use it extensively for testing and in external_question_result to give the result
+        // variable a default value (we could probably fix this some other way).
+        // If we don't disable rebinding, we need to do something with the rebind_variable_group
+        // test so that the behavior is better defined.
+        // assert!(!matches!(self.variable_state(var), VariableState::Bound(_)), "Variable is bound");
+
+        if let Ok(symbol) = val.value().as_symbol() {
+            self.bind_variables(var, symbol);
+        } else {
+            self.add_binding(var, val);
+        }
+    }
+
+    /// Bind two variables together.
+    fn bind_variables(&mut self, left: &Symbol, right: &Symbol) {
+        match (self.variable_state(left), self.variable_state(right)) {
+            (VariableState::Bound(_), VariableState::Unbound) => {
```

Comment:
Oh, these are generally being handled in `unify_vars`, right?

Reply:
Yes, all used to be in unify vars. Alex removed here https://github.com/osohq/oso/pull/723/commits/ba82907976a42e9baf15a6e1a37d905d86c222b3